File: src/main/java/work/alsace/buildercore/listeners/PlayerListener.java
package work.alsace.buildercore.listeners;

import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.FoodLevelChangeEvent;
import org.bukkit.event.inventory.PrepareAnvilEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerKickEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerTeleportEvent;
import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import work.alsace.buildercore.BuilderCore;
import work.alsace.buildercore.Utils.NoClipUtil;
import work.alsace.buildercore.commands.builderTools.AdvanceFlyCommand;
import work.alsace.buildercore.commands.teleport.BackCommand;
import work.alsace.buildercore.service.HomeDataLoader;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class PlayerListener implements Listener {
    private final BuilderCore plugin;

    public PlayerListener(BuilderCore plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        event.setJoinMessage(null);
        BlockListener.slabs.add(event.getPlayer());
        plugin.getHomeProfiles().put(event.getPlayer().getUniqueId(), new HomeDataLoader(event.getPlayer().getUniqueId(), plugin));
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        event.setQuitMessage(null);
        this.plugin.getHasIgnored().remove(event.getPlayer().getName());
        plugin.getHomeProfiles().remove(event.getPlayer().getUniqueId());
        if (NoClipUtil.noclip.contains(event.getPlayer()) || AdvanceFlyCommand.enabledPlayers.contains(event.getPlayer()) || BlockListener.slabs.contains(event.getPlayer())) {
            NoClipUtil.noclip.remove(event.getPlayer());
            AdvanceFlyCommand.enabledPlayers.remove(event.getPlayer());
            BlockListener.slabs.remove(event.getPlayer());
        }
        event.getPlayer().removeMetadata("afk", plugin);
    }

    @EventHandler
    public void onPlayerTeleport(PlayerTeleportEvent event) {
        if (event.getCause().equals(TeleportCause.SPECTATE) && !event.getPlayer().hasPermission("buildercore.event.spectate")) {
            event.setCancelled(true);
        }

        // 检查/back命令是否已注册
        if (plugin.getCommand("back") != null) {
            BackCommand backCommand = (BackCommand) Objects.requireNonNull(plugin.getCommand("back")).getExecutor();
            backCommand.addToHistory(event.getPlayer());
        }
    }

    @EventHandler
    public void onPlayerKick(PlayerKickEvent event) {
        plugin.getHomeProfiles().remove(event.getPlayer().getUniqueId());
    }

    @EventHandler
    public void playerAnvil(PrepareAnvilEvent event) {
        ItemStack item = event.getResult();
        if (item != null) {
            ItemMeta meta = item.getItemMeta();
            if (meta != null && meta.hasDisplayName()) {
                String name = meta.getDisplayName();
                name = ChatColor.translateAlternateColorCodes('&', name);
                meta.setDisplayName(this.translateHexColorCodes(name));
                item.setItemMeta(meta);
                event.setResult(item);
            }
        }
    }

    @EventHandler
    public void onPlayerDamage(EntityDamageEvent event) {
        if (event.getEntity() instanceof Player) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onPlayerFoodLevelChange(FoodLevelChangeEvent event) {
        if (event.getEntity() instanceof Player) {
            event.setCancelled(true);
        }
    }

    private String translateHexColorCodes(String message) {
        Pattern hexPattern = Pattern.compile("#([A-Fa-f0-9]{6})");
        Matcher matcher = hexPattern.matcher(message);
        StringBuilder buffer = new StringBuilder(message.length() + 32);

        while (matcher.find()) {
            String group = matcher.group(1);
            char var10002 = group.charAt(0);
            matcher.appendReplacement(buffer, "§x§" + var10002 + "§" + group.charAt(1) + "§" + group.charAt(2) + "§" + group.charAt(3) + "§" + group.charAt(4) + "§" + group.charAt(5));
        }

        return matcher.appendTail(buffer).toString();
    }
}

File: src/main/java/work/alsace/buildercore/listeners/PlayerListener.kt
package work.alsace.buildercore.listeners

import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.FoodLevelChangeEvent
import org.bukkit.event.inventory.PrepareAnvilEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerKickEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.player.PlayerTeleportEvent
import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause
import work.alsace.buildercore.BuilderCore
import work.alsace.buildercore.commands.builderTools.AdvanceFlyCommand
import work.alsace.buildercore.commands.teleport.BackCommand
import work.alsace.buildercore.service.HomeDataLoader
import work.alsace.buildercore.utils.NoClipUtil
import java.util.*
import java.util.regex.Pattern

class PlayerListener(private val plugin: BuilderCore) : Listener {
    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        event.joinMessage = null
        BlockListener.slabs.add(event.player)
        plugin.homeProfiles[event.player.uniqueId] = HomeDataLoader(event.player.uniqueId, plugin)
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        event.quitMessage = null
        plugin.hasIgnored.remove(event.player.name)
        plugin.homeProfiles.remove(event.player.uniqueId)
        if (NoClipUtil.noclip.contains(event.player) || AdvanceFlyCommand.enabledPlayers.contains(
                event.player
            ) || BlockListener.slabs.contains(event.player)
        ) {
            NoClipUtil.noclip.remove(event.player)
            AdvanceFlyCommand.enabledPlayers.remove(event.player)
            BlockListener.slabs.remove(event.player)
        }
        event.player.removeMetadata("afk", plugin)
    }

    @EventHandler
    fun onPlayerTeleport(event: PlayerTeleportEvent) {
        if (event.cause == TeleportCause.SPECTATE && !event.player.hasPermission("buildercore.event.spectate")) {
            event.isCancelled = true
        }

        // 检查/back命令是否已注册
        if (plugin.getCommand("back") != null) {
            val backCommand = Objects.requireNonNull(plugin.getCommand("back"))?.executor as BackCommand
            backCommand.addToHistory(event.player)
        }
    }

    @EventHandler
    fun onPlayerKick(event: PlayerKickEvent) {
        plugin.homeProfiles.remove(event.player.uniqueId)
    }

    @EventHandler
    fun playerAnvil(event: PrepareAnvilEvent) {
        val item = event.result
        if (item != null) {
            val meta = item.itemMeta
            if (meta != null && meta.hasDisplayName()) {
                var name = meta.displayName
                name = ChatColor.translateAlternateColorCodes('&', name)
                meta.setDisplayName(translateHexColorCodes(name))
                item.setItemMeta(meta)
                event.result = item
            }
        }
    }

    @EventHandler
    fun onPlayerDamage(event: EntityDamageEvent) {
        if (event.entity is Player) {
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onPlayerFoodLevelChange(event: FoodLevelChangeEvent) {
        if (event.entity is Player) {
            event.isCancelled = true
        }
    }

    private fun translateHexColorCodes(message: String): String {
        val hexPattern = Pattern.compile("#([A-Fa-f0-9]{6})")
        val matcher = hexPattern.matcher(message)
        val buffer = StringBuilder(message.length + 32)
        while (matcher.find()) {
            val group = matcher.group(1)
            val var10002 = group[0]
            matcher.appendReplacement(
                buffer,
                "§x§" + var10002 + "§" + group[1] + "§" + group[2] + "§" + group[3] + "§" + group[4] + "§" + group[5]
            )
        }
        return matcher.appendTail(buffer).toString()
    }
}
